### ISSUE TITLE
On usability of NPM and JSPM manager 

### ISSUE THREAD

**Issue by adriatic** [monterey-framework/monterey#360]
### Intention

**I'm initiating a discussion that should result with  a feature request**

**Comment by adriatic** [comment]
### Original context from Gitter chat
![image](https://cloud.githubusercontent.com/assets/2712405/19040132/d80e90f8-8950-11e6-9999-813eeef1e800.png)

**Comment by adriatic** [comment]
My explanations to items tagged with red markers:

### Marker 1
Before making any "sweeping" change (like upgrading the Aurelia framework version) Monterey needs to make backup copies of all affected files (`package.json` / `config.js` when using SystemJS module loader) to allow eventual "undo" later.

### Marker 2
This is a complex way of saying that I would like to see Monterey know all versions of all modules of interest and be able to find "fitting combinations". 

**Example**: 
command `npm view aurelia-kendoui-bridge versions` will list all registered versions of aurelia-kendoui-bridge:

![image](https://cloud.githubusercontent.com/assets/2712405/19041564/906e56d8-8956-11e6-8c8b-020988c3d407.png)

`npm view aurelia-templating versions` yields

![image](https://cloud.githubusercontent.com/assets/2712405/19041540/7cf06db2-8956-11e6-9afe-23dad2fe32fc.png)

Monterey user should be able to say I want to use the latest version of KendoUI bridge that works with Aurelia framework that has its ` aurelia-templating ver. 1.0.0-rc.1.0.1`

Anyone claiming how this is an unreasonable expectation, imagine the example that is closest to this one and is condered reasonable 😏 )

### Marker 3

### Marker 4.

**Comment by JeroenVinke** [comment]
>I may want to upgrade to Aurelia version "one before latest" for example

This is **exactly** something I wanted to add to the JSPM/NPM package manager screens.

In https://github.com/monterey-framework/monterey/issues/99#issuecomment-239767318 I wrote this:

> Right now we have to do tens of requests to NPM/Github to find out the latest version of a package. There isn't a nice http endpoint where we can say "Get me the versions of packages X, Y and Z" with a single http request. Right now we need to do a requests for X, Y an Z seperately. If we build a http API then monterey can do a single request, and the http api can do caching to speed up the process even more.

Maybe there are existing solutions for this already, but this is the problem we need to solve before we can implement the feature that you are proposing. I have opened up https://github.com/monterey-framework/monterey/issues/361 for when we don't find existing solutions

> Monterey user should be able to say I want to use the latest version of KendoUI bridge that works with Aurelia framework that has its aurelia-templating ver. 1.0.0-rc.1.0.1

What you are explaining here is semver, because based on semver you should be able to figure out compatibility between packages. JSPM is doing this when we call `jspm install` or `jspm update`

**Comment by JeroenVinke** [comment]
A different feature, similar to this, would be to remove (and add again) the caret (`^`) sign in front of all dependencies of your app, so that JSPM will not automatically download the latest version of a library without you knowing

**Comment by JeroenVinke** [comment]
And i'd like to have some sort of fork analyzer, something that could help you figure out why the hell forks were created. Right now we this in a very basic form (it will warn you that there are forks, and what packages were forked) but not why

**Comment by adriatic** [comment]
I may have not been sufficiently clear - my expected solution requires more than `semver` which I saw failing often. I used the term **expert system** defined as

> In artificial intelligence, an expert system is a computer system that emulates the decision-making ability of a human expert. Expert systems are designed to solve complex problems by reasoning about knowledge, represented primarily as if–then rules rather than through conventional procedural code

So, our database would be peppered with `if-then-rules` that the whole community maintains (with supervision of course).

Again, as before having this expert system in our plans does not preclude **any** of the solutions being discussed in this thread. 

My dream goal would be a query like: 'find the best combination of aurelia-framework 1.0.2 with kendoui-bridge and jQuery newer than 2.4` - and that comes back with `done, your old package.json / config.js saved -- click here to revert`

**Comment by adriatic** [comment]
Love the `discussion` label!!

**Comment by adriatic** [comment]
A triple **`YESSSS`** to the **fork analyzer](https://github.com/monterey-framework/monterey/issues/360#issuecomment-251137132)** idea

**Comment by JeroenVinke** [comment]
> So, our database would be peppered with if-then-rules that the whole community maintains (with supervision of course).

The problem I see here is that there are so many different packages, each releasing new versions pretty rapidly, so the chance of Monterey knowing about one particular compatibility issue between two packages is very small. Also, little reward of reporting compatibility issues (you report something but will probably never hear back) could be an issue as well

**Comment by adriatic** [comment]
I completely agree with this - which is why I am defining such unlikely to reach goal. My hope is that might use some of the tricks from the that "expert systems bag"
